js/UrbanAirship.js

```diff
@@ -365,9 +368,12 @@ class UrbanAirship {
    *
    * @param {string} eventName The event name. Either notificationResponse, pushReceived,
    * register, deepLink, or notificationOptInStatus.
-   * @param {Function} listener The event listner.
+   * @param {Function} listener The event listener.
    */
   static removeListener(eventName: AirshipEventName, listener: Function) {
+    if (Platform.OS === 'android') {
+      UrbanAirshipModule.decreaseListenerCount();
```

Comment:
This may cause some issues if the subscription was cancelled from the return type of addListener of if removeListener is called multiple times with the same values. We should probably extend the class like react does and decrease the count in removeSubscription - https://github.com/facebook/react-native/blob/a93b7a2da0849f15708afef1ff99022d8cc55b14/Libraries/EventEmitter/NativeEventEmitter.js#L53

Reply:
I agree that what you suggest is a more correct approach. But it would mean building a custom `NativeEventEmitter`, since extending the existing one wouldn't do the job. This is because React Native 0.51.0 [introduced some changes](https://github.com/facebook/react-native/commit/2fff445b13da114cc5238ad9064ed2100030d5b9#diff-cb16de13bba13fdbc13d117d51774caf) in the `NativeEventEmitter` class. Before 0.51.0 `_nativeModule.removeListener` was only triggered on iOS. Now, with this new version, it would also be triggered on Android. So this custom event emitter should probably have to extend [`EventEmitter` itself](https://github.com/facebook/react-native/blob/master/Libraries/vendor/emitter/EventEmitter.js) in order to avoid this inconsistency.

Anyhow, in this particular case, where it is used only for `increaseListenerCount` / `decreaseListenerCount`, what I coded shouldn't lead to any major issues. I tested it on our app, where I can reproduce the [issue I reported](https://github.com/urbanairship/react-native-module/issues/58), and I can confirm that this PR fixes it.

Reply:
I see. So we either have to force react > 0.51 or build our own. The react extension is pretty minor, providing our own is not bad:

```
// @flow
'use strict';

const EventEmitter = require('EventEmitter');
const RCTDeviceEventEmitter = require('RCTDeviceEventEmitter');

const invariant = require('fbjs/lib/invariant');

import type EmitterSubscription from 'EmitterSubscription';

type NativeModule = {
  +addListener: (eventType: string) => void,
  +removeListeners: (count: number) => void,
};

class UrbanAirshipEventEmitter extends EventEmitter {
  _nativeModule: ?NativeModule;

  constructor(nativeModule: ?NativeModule) {
    super(RCTDeviceEventEmitter.sharedSubscriber);
    invariant(nativeModule, 'Native module cannot be null.');
    this._nativeModule = nativeModule;
  }

  addListener(eventType: string, listener: Function, context: ?Object): EmitterSubscription {
    this._nativeModule.addListener(eventType);
    return super.addListener(eventType, listener, context);
  }

  removeAllListeners(eventType: string) {
    invariant(eventType, 'eventType argument is required.');
    const count = this.listeners(eventType).length;
    this._nativeModule.removeListeners(count);
    super.removeAllListeners(eventType);
  }

  removeSubscription(subscription: EmitterSubscription) {
    this._nativeModule.removeListeners(1);
    super.removeSubscription(subscription);
  }
}

module.exports = UrbanAirshipEventEmitter;
```

Then we just need to update the methods you provided to be `removeListeners` and `addListener`